jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/AbstractDeploymentService.java

```diff
@@ -176,19 +179,44 @@ public void commonDeploy(DeploymentUnit unit, DeployedUnitImpl deployedUnit, Run
     
     @Override
     public void undeploy(DeploymentUnit unit) {
+        undeploy(unit, false);
+    }
+
+    @Override
+    public void undeploy(DeploymentUnit unit, boolean abortInstances) {
+
         List<Integer> states = new ArrayList<Integer>();
         states.add(ProcessInstance.STATE_ACTIVE);
         states.add(ProcessInstance.STATE_PENDING);
         states.add(ProcessInstance.STATE_SUSPENDED);
         Collection<ProcessInstanceDesc> activeProcesses = runtimeDataService.getProcessInstancesByDeploymentId(unit.getIdentifier(), states, new QueryContext());
-        if (!activeProcesses.isEmpty()) {
-            throw new IllegalStateException("Undeploy forbidden - there are active processes instances for deployment " 
-                                            + unit.getIdentifier());
+        if (!abortInstances && !activeProcesses.isEmpty()) {
+            throw new IllegalStateException("Undeploy forbidden - there are active processes instances for deployment "
+                                                    + unit.getIdentifier());
         }
+
         synchronized (this) {
             DeployedUnit deployed = deploymentsMap.remove(unit.getIdentifier());
+
             if (deployed != null) {
+
                 RuntimeManager manager = deployed.getRuntimeManager();
+
+                if (abortInstances && !activeProcesses.isEmpty()) {
+                    activeProcesses = runtimeDataService.getProcessInstancesByDeploymentId(unit.getIdentifier(), states, new QueryContext(0, -1));
+                    activeProcesses.forEach(processInstanceDesc -> {
+                        RuntimeEngine engine = manager.getRuntimeEngine(ProcessInstanceIdContext.get(processInstanceDesc.getId()));
+                        try {
+                            KieSession ksession = engine.getKieSession();
+                            ksession.abortProcessInstance(processInstanceDesc.getId());
```

Comment:
@mswiderski we should reconsider throwing this Exception, if we throw it it won't be possible to update a snapshot without aborting process instances. WDYT?

Reply:
maybe we can add an extra param, to skip the check.

Reply:
I see your point, instead of adding another parameter maybe we could use Function<Collection<ProcessInstanceDesc>, Boolean> that would be responsible for dealing with these checks and operations. We could then have few implementation
- default that checks if the list is empty and if so disallow underlay
- one that aborts all process instances from the list and if all were done successfully returns true to proceed with undeploy
- another one that simply ignores the list of instances and ask to underplay anyway

then Kie Server based on all the info it has can select one of them to make the undeploy/update/etc

wdyt?

Reply:
Probably what's the best solution. What do you think about instead of using a Function use a Function<DeploymentUnit, Boolean> and let the consumer be the one that gets the active instances and decides what to do, there might be cases that we don't need any check and we can just skip the query.

Reply:
@mswiderski ^^

Reply:
maybe it is more concise but it does not reflect the need as it is expected to perform actions inside it. So fine to use DeploymentUnit to avoid extra query) but we should not use Predicate

Reply:
@mswiderski +1